src/lib/modernExtend.ts

```diff
@@ -67,13 +67,15 @@ async function setupAttributes(
 
 export function setupConfigureForReporting(
     cluster: string | number, attribute: ReportingConfigAttribute, config: ReportingConfigWithoutAttribute, access: Access,
+    endpointID?: number,
```

Comment:
Why is this needed? `setupAttributes` automatically check the clusters.

Reply:
It is true, but if device has, say, 10 endpoints with temperature cluster - then `setupAttributes` function currently will iterate all endpoints and for each endpoint with temperature cluster it will try to bind it, setup reporting and possibly read the value of the attribute. As temperature cluster is present in 10 endpoints - `setupAttributes` will be called 10 times. This results in 10^2 * 3 = 300 operations.

If instead endpoint will be provided - `setupAttributes` will only do necessary operations for each endpoint and cluster exactly once, reducing operations to 10 * 3 = 30.

Does this sound reasonable?

Reply:
But why are there 10 endpoints with the temperature cluster on the device?

Reply:
Just an arbitrary example.

Still, it is possible that device could have more than one endpoint with same cluster. And if it would be cluster like temperature or pressure(or any numeric one) - would trigger this behavior.

As another example: device has multiple environmental sensors attached, and those sensors have overlapping and distinct clusters. I.e. all sensors can have temperature cluster, but only one of them will have CO2, another one can have VOC, etc. Think like board can have attached Bosch BME280 & Sensirion SCD4X sensors. They both have temp & humidity clusters, but BME280 additionally has pressure, while SCD4X has CO2.

Reply:
If I understand correctly, for your use-case it's important that the generate definition works without requiring any external converters. How would the generate definition decide for which endpoints to setup temperature reporting for if e.g. 10 of the endpoints have the temperature cluster? It sounds like we are fixing something in z2m here while it should be fixed at the device side instead.

Reply:
> How would the generate definition decide for which endpoints to setup temperature reporting for if e.g. 10 of the endpoints have the temperature cluster?

If we provide specific endpoint we can then tailor extender for that specific cluster & endpoint(would it be cluster attribute value or endpoint to bind to).

> It sounds like we are fixing something in z2m here while it should be fixed at the device side instead.

Maybe we have a bit of misunderstanding.
For my example I chose temp cluster, because I tested with it, but consider other examples of clusters that also rely on this functionality for generated definition: 
* [onOff](https://github.com/Koenkk/zigbee-herdsman-converters/blob/6a0797ab702db6d92113bd1389f1ca29933c4ede/src/lib/modernExtend.ts#L117)
* [light](https://github.com/Koenkk/zigbee-herdsman-converters/blob/6a0797ab702db6d92113bd1389f1ca29933c4ede/src/lib/modernExtend.ts#L313)
* [lock](https://github.com/Koenkk/zigbee-herdsman-converters/blob/6a0797ab702db6d92113bd1389f1ca29933c4ede/src/lib/modernExtend.ts#L355)
* [electricityMeter](https://github.com/Koenkk/zigbee-herdsman-converters/blob/6a0797ab702db6d92113bd1389f1ca29933c4ede/src/lib/modernExtend.ts#L226)

I am suggesting that device can have multiple endpoints with the same cluster, and we should try to support it by generated definitions as well.
For example devices that have multiple endpoints with same cluster:
* [FUGA compatible switch from Smarthjemmet.dk](https://www.zigbee2mqtt.io/devices/QUAD-ZIG-SW.html)
* [Namron 4512768](https://www.zigbee2mqtt.io/devices/4512768.html)
* [Aurora Lighting AU-A1ZBDSS](https://www.zigbee2mqtt.io/devices/AU-A1ZBDSS.html)

Reply:
> I am suggesting that device can have multiple endpoints with the same cluster, and we should try to support it by generated definitions as well.

Yes that would be nice, but instead of having e.g. `temperature(endpointID: 1)`, `temperatature(endpointID: 2`), we can do something similar to what we already do for [onOff](https://github.com/Koenkk/zigbee-herdsman-converters/blob/6a0797ab702db6d92113bd1389f1ca29933c4ede/src/devices/atsmart.ts#L10). The only thing that needs to be added multiple times is the expose, to/fromZigbee converters don't need to be duplicated.

Reply:
Ah, I see.
Thank you for explanation, will uldate the implementation then!

Reply:
Sorry, took a bit longer then I expected. Now `from` and `to` entities are not duplicated.